internal/querycoordv2/meta/target_manager.go

```diff
@@ -1,186 +1,203 @@
+		for _, segmentBinlog := range binlogs {
+			mgr.Next.addSegment(&datapb.SegmentInfo{
+				ID:            segmentBinlog.GetSegmentID(),
+				CollectionID:  collectionID,
+				PartitionID:   partitionID,
+				InsertChannel: segmentBinlog.GetInsertChannel(),
+				NumOfRows:     segmentBinlog.GetNumOfRows(),
+				Binlogs:       segmentBinlog.GetFieldBinlogs(),
+				Statslogs:     segmentBinlog.GetStatslogs(),
+				Deltalogs:     segmentBinlog.GetDeltalogs(),
+			})
 		}
 	}
+
+	log.Info("UpdateNextTarget finish", zap.Int64("collectionID", collectionID),
+		zap.Int64s("partitionIDs", partitionIDs))
 	return nil
 }
 
-func (mgr *TargetManager) ContainDmChannel(channel string) bool {
-	mgr.rwmutex.RLock()
-	defer mgr.rwmutex.RUnlock()
+func (mgr *TargetManager) mergeDmChannelInfo(infos []*datapb.VchannelInfo) *DmChannel {
+	var dmChannel *DmChannel
 
-	_, ok := mgr.dmChannels[channel]
-	return ok
-}
+	for _, info := range infos {
+		if dmChannel == nil {
+			dmChannel = DmChannelFromVChannel(info)
+			continue
+		}
+
+		if info.SeekPosition.GetTimestamp() < dmChannel.SeekPosition.GetTimestamp() {
+			dmChannel.SeekPosition = info.SeekPosition
+		}
+		dmChannel.DroppedSegmentIds = append(dmChannel.DroppedSegmentIds, info.DroppedSegmentIds...)
+		dmChannel.UnflushedSegmentIds = append(dmChannel.UnflushedSegmentIds, info.UnflushedSegmentIds...)
+		dmChannel.FlushedSegmentIds = append(dmChannel.FlushedSegmentIds, info.FlushedSegmentIds...)
+	}
 
-func (mgr *TargetManager) removeDmChannel(channel string) {
-	delete(mgr.dmChannels, channel)
-	log.Info("remove channel from targets", zap.String("channel", channel))
+	return dmChannel
 }
 
-func (mgr *TargetManager) GetDmChannelsByCollection(collectionID int64) []*DmChannel {
-	mgr.rwmutex.RLock()
-	defer mgr.rwmutex.RUnlock()
+// RemoveCollection removes all channels and segments in the given collection
+func (mgr *TargetManager) RemoveCollection(collectionID int64) {
+	mgr.Current.RWMutex.Lock()
+	defer mgr.Current.RWMutex.Unlock()
+	mgr.Next.RWMutex.Lock()
+	defer mgr.Next.RWMutex.Unlock()
 
-	channels := make([]*DmChannel, 0)
-	for _, channel := range mgr.dmChannels {
-		if channel.GetCollectionID() == collectionID {
-			channels = append(channels, channel)
-		}
-	}
-	return channels
+	log.Info("remove collection from current target", zap.Int64("collectionID", collectionID))
+	mgr.Current.removeCollection(collectionID)
+
+	log.Info("remove collection from Next target", zap.Int64("collectionID", collectionID))
+	mgr.Next.removeCollection(collectionID)
 }
 
-func (mgr *TargetManager) GetSegment(id int64) *datapb.SegmentInfo {
-	mgr.rwmutex.RLock()
-	defer mgr.rwmutex.RUnlock()
+// RemovePartition removes all segment in the given partition,
+// NOTE: this doesn't remove any channel even the given one is the only partition
+func (mgr *TargetManager) RemovePartition(partitionID int64) {
+	mgr.Current.RWMutex.Lock()
+	defer mgr.Current.RWMutex.Unlock()
+	mgr.Next.RWMutex.Lock()
+	defer mgr.Next.RWMutex.Unlock()
```

Comment:
Generally, a lock of `TargetManager` is better than acquiring 2 locks of `Target`